Simplify serialization of ObjCObjectGraph

csaavedra opened the pull request:
#### e1cb46765c090e65d6ad2a70b32263699d683d92
<pre>
Simplify serialization of ObjCObjectGraph
<a href="https://bugs.webkit.org/show_bug.cgi?id=272585">https://bugs.webkit.org/show_bug.cgi?id=272585</a>

Reviewed by Darin Adler.

NSDictionary and NSArray already have serialization support,
so there is no need to cook it by hand.

* Source/WebKit/Shared/mac/ObjCObjectGraph.mm:
(WebKit::ObjCObjectGraph::encode):
(WebKit::ObjCObjectGraph::decode):

Canonical link: <a href="https://commits.webkit.org/277454@main">https://commits.webkit.org/277454@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f75afd0f37923b890f772e903c8e36768dde628e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47649 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26837 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50326 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50332 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43704 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32675 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24302 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38792 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48231 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24487 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41061 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20089 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21970 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5698 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43992 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42668 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52224 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22685 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19012 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46096 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23957 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45131 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24745 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6735 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23677 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/Shared/mac/ObjCObjectGraph.mm:
```diff
@@ -148,11 +148,7 @@ static bool shouldTransformGraph(id object, const ObjCObjectGraph::Transformer&
-        encoder << static_cast<uint64_t>(array.count);
-        for (id element in array)
-            encode(encoder, element);
+        IPC::encode(encoder, static_cast<NSArray *>(object));
```
Should be `checked_objc_cast` rather than `static_cast`.

csaavedra replied:
Right. The cases handling other types in the same method are all using `static_cast<>()` though, which is why I did the same. Should we convert them too to use a checked ObjectC cast?

csaavedra replied:
On second thought, I think that that the logic here is that there is already certainty that the type being casted to is correct, because `typeFromObject()`, used to find out the type of the object, is already using `dynamic_objc_cast<>()`. The checked cast would be redundant, since we are certain that the type is correct, so I will leave the static cast.

darinadler replied:
I am going to change these _all_ to `checked_objc_cast` in a patch later. Your logic about why we don’t need a check applies to _all_ places we use `checked_objc_cast` already. The point is to add an extra check in case our thinking is wrong, to harden for security purposes.

csaavedra replied:
Sure, thank you!

csaavedra replied:
Actually I took the freedom to submit a PR already, feel free to review. Thanks again! https://github.com/WebKit/WebKit/pull/27219

darinadler commented on Source/WebKit/Shared/mac/ObjCObjectGraph.mm:
```diff
@@ -167,13 +163,7 @@ static bool shouldTransformGraph(id object, const ObjCObjectGraph::Transformer&
-            encode(encoder, key);
-            encode(encoder, object);
-        }];
+        IPC::encode(encoder, static_cast<NSDictionary *>(object));
```
Should be `checked_objc_cast` rather than `static_cast`.

darinadler commented on Source/WebKit/Shared/mac/ObjCObjectGraph.mm:
```diff
@@ -210,19 +200,11 @@ static bool shouldTransformGraph(id object, const ObjCObjectGraph::Transformer&
     case ObjCType::NSArray: {
-        uint64_t size;
-        if (!decoder.decode(size))
+        std::optional<RetainPtr<NSArray>> array = decoder.decode<RetainPtr<NSArray>>();
```
Can we use `auto` here instead?

csaavedra replied:
Right. I will update the patch to use `auto` in all cases.

darinadler commented on Source/WebKit/Shared/mac/ObjCObjectGraph.mm:
```diff
@@ -245,28 +227,11 @@ static bool shouldTransformGraph(id object, const ObjCObjectGraph::Transformer&
     case ObjCType::NSDictionary: {
-        uint64_t size;
-        if (!decoder.decode(size))
+        std::optional<RetainPtr<NSDictionary>> dictionary = decoder.decode<RetainPtr<NSDictionary>>();
```
Can we use `auto` here instead?

csaavedra committed:
Simplify serialization of ObjCObjectGraph
https://bugs.webkit.org/show_bug.cgi?id=272585

Reviewed by Darin Adler.

NSDictionary and NSArray already have serialization support,
so there is no need to cook it by hand.

* Source/WebKit/Shared/mac/ObjCObjectGraph.mm:
(WebKit::ObjCObjectGraph::encode):
(WebKit::ObjCObjectGraph::decode):

Canonical link: https://commits.webkit.org/277454@main
